dask_cuda/proxify_host_file.py

```diff
@@ -288,20 +373,47 @@ class ProxifyHostFile(MutableMapping):
     ----------
     device_memory_limit: int
         Number of bytes of CUDA device memory used before spilling to host.
-    compatibility_mode: bool or None
+    host_memory_limit: int
+        Number of bytes of host memory used before spilling to disk.
+    local_directory: str or None, default None
+        Path on local machine to store temporary files. Can be a string (like
+        ``"path/to/files"``) or ``None`` to fall back on the value of
+        ``dask.temporary-directory`` in the local Dask configuration, using the
+        current working directory if this is not set.
+        WARNING, this **cannot** change while running thus all serialization to
+        disk are using the same directory.
+    shared_filesystem: bool or None, default None
+        Whether the `local_directory` above is shared between all workers or not.
+        If ``None``, the "jit-unspill-shared-fs" config value are used, which
+        defaults to False.
+        Notice, a shared filesystem must support the `os.link()` operation.
+    compatibility_mode: bool or None, default None
         Enables compatibility-mode, which means that items are un-proxified before
         retrieval. This makes it possible to get some of the JIT-unspill benefits
         without having to be ProxyObject compatible. In order to still allow specific
         ProxyObjects, set the `mark_as_explicit_proxies=True` when proxifying with
-        `proxify_device_objects()`. If None, the "jit-unspill-compatibility-mode"
+        `proxify_device_objects()`. If ``None``, the "jit-unspill-compatibility-mode"
         config value are used, which defaults to False.
     """
 
-    def __init__(self, device_memory_limit: int, compatibility_mode: bool = None):
-        self.device_memory_limit = device_memory_limit
+    _spill_directory: Optional[str] = None
+    _spill_shared_filesystem: bool
+    _spill_to_disk_prefix: str = f"spilled-data-{uuid.uuid4()}"
+    _spill_to_disk_counter: int = 0
+    lock = threading.RLock()
```

Comment:
Why not define/initialize these attributes in `__init__`?

Reply:
Looking later on in the class, it seems that this is because you have multiple `@classmethod`s, is that right? In that case, why can't all of them call `__init__` (i.e., `cls(...)`)? The way this code is written looks confusing to me, it's not really clear when `__init__` will be called and for what purposes it should or shouldn't be called. IOW, I would expect `__init__` to be always called and have some common code to all possible initializers, leaving `@classmethod`s as the only way to create the object, or possibly allow `Class()` with some simple/general behavior too.

Reply:
It is because `register_serialization_family()` is a global Dask operation that we call once. I have added the follow to make it more clear:
```python
    @classmethod
    def register_disk_spilling(
        cls, local_directory: str = None, shared_filesystem: bool = None
    ):
        """Register Dask serializers that writes to disk

        This is a static method because the registration of a Dask
        serializer/deserializer par is a global operation thus we can
        only register one such par. This means that all instances of
        the ProxifyHostFile end up using the same local_directory.
        """
```
Notice, `__init__` is called once per worker, which translate into one call per process in Dask-CUDA since we spawn exactly one worker per process per GPU. However, in principle we could spawn multiple workers per process, which this approach support.